# Assignment

In [1]:
from warnings import filterwarnings
filterwarnings('ignore')

### Reading the Dataset

In [2]:
import os
os.chdir('D:/Dataset/')

In [3]:
import pandas as pd
df = pd.read_csv('training_set.csv')
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

In [5]:
s = df.isna().sum()
s

Id                 0
MSSubClass         0
MSZoning           0
LotFrontage      259
LotArea            0
                ... 
MoSold             0
YrSold             0
SaleType           0
SaleCondition      0
SalePrice          0
Length: 81, dtype: int64

In [6]:
s[s>0]

LotFrontage      259
Alley           1369
MasVnrType         8
MasVnrArea         8
BsmtQual          37
BsmtCond          37
BsmtExposure      38
BsmtFinType1      37
BsmtFinType2      38
Electrical         1
FireplaceQu      690
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
PoolQC          1453
Fence           1179
MiscFeature     1406
dtype: int64

### Seperate X and Y

In [7]:
X = df.drop(columns=['Id','SalePrice'])
Y = df[['SalePrice']]

In [8]:
X.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          60       RL         65.0     8450   Pave   NaN      Reg   
1          20       RL         80.0     9600   Pave   NaN      Reg   
2          60       RL         68.0    11250   Pave   NaN      IR1   
3          70       RL         60.0     9550   Pave   NaN      IR1   
4          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig  ... ScreenPorch PoolArea PoolQC Fence  \
0         Lvl    AllPub    Inside  ...           0        0    NaN   NaN   
1         Lvl    AllPub       FR2  ...           0        0    NaN   NaN   
2         Lvl    AllPub    Inside  ...           0        0    NaN   NaN   
3         Lvl    AllPub    Corner  ...           0        0    NaN   NaN   
4         Lvl    AllPub       FR2  ...           0        0    NaN   NaN   

  MiscFeature MiscVal  MoSold  YrSold  SaleType  SaleCondition  
0         NaN       0       2    2008        WD         Normal  
1         NaN       0       5    2007        WD         Normal  
2         NaN       0       9    2008        WD         Normal  
3         NaN       0       2    2006        WD        Abnorml  
4         NaN       0      12    2008        WD         Normal  

[5 rows x 79 columns]

In [9]:
Y.head()

SalePrice
0     208500
1     181500
2     223500
3     140000
4     250000

### Cat con seperation

In [10]:
from PM6 import catconsep

In [11]:
cat, con = catconsep(X)

In [12]:
cat

['MSZoning',
 'Street',
 'Alley',
 'LotShape',
 'LandContour',
 'Utilities',
 'LotConfig',
 'LandSlope',
 'Neighborhood',
 'Condition1',
 'Condition2',
 'BldgType',
 'HouseStyle',
 'RoofStyle',
 'RoofMatl',
 'Exterior1st',
 'Exterior2nd',
 'MasVnrType',
 'ExterQual',
 'ExterCond',
 'Foundation',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'BsmtFinType1',
 'BsmtFinType2',
 'Heating',
 'HeatingQC',
 'CentralAir',
 'Electrical',
 'KitchenQual',
 'Functional',
 'FireplaceQu',
 'GarageType',
 'GarageFinish',
 'GarageQual',
 'GarageCond',
 'PavedDrive',
 'PoolQC',
 'Fence',
 'MiscFeature',
 'SaleType',
 'SaleCondition']

In [13]:
con

['MSSubClass',
 'LotFrontage',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'YearRemodAdd',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 '1stFlrSF',
 '2ndFlrSF',
 'LowQualFinSF',
 'GrLivArea',
 'BsmtFullBath',
 'BsmtHalfBath',
 'FullBath',
 'HalfBath',
 'BedroomAbvGr',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageYrBlt',
 'GarageCars',
 'GarageArea',
 'WoodDeckSF',
 'OpenPorchSF',
 'EnclosedPorch',
 '3SsnPorch',
 'ScreenPorch',
 'PoolArea',
 'MiscVal',
 'MoSold',
 'YrSold']

### Start feature selection pipeline
Categorical Feature : Ordinal Encoder + StandardScaler

In [14]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OrdinalEncoder, StandardScaler
from sklearn.compose import ColumnTransformer

In [15]:
num_pipe1 = Pipeline(steps=[('imputer',SimpleImputer(strategy='mean')),
                            ('scaler',StandardScaler())])

cat_pipe1 = Pipeline(steps=[('imputer',SimpleImputer(strategy='constant', fill_value='NotAvailable')),
                            ('ordinal',OrdinalEncoder()),
                            ('scaler',StandardScaler())])

pre1 = ColumnTransformer([('num',num_pipe1, con),
                          ('cat',cat_pipe1, cat)])

In [16]:
X_pre = pre1.fit_transform(X)
X_pre

array([[ 0.07337496, -0.22937175, -0.20714171, ..., -0.1859753 ,
         0.31386709,  0.2085023 ],
       [-0.87256276,  0.4519361 , -0.09188637, ..., -0.1859753 ,
         0.31386709,  0.2085023 ],
       [ 0.07337496, -0.09311018,  0.07347998, ..., -0.1859753 ,
         0.31386709,  0.2085023 ],
       ...,
       [ 0.30985939, -0.18395123, -0.14781027, ...,  5.19073639,
         0.31386709,  0.2085023 ],
       [-0.87256276, -0.09311018, -0.08016039, ..., -0.1859753 ,
         0.31386709,  0.2085023 ],
       [-0.87256276,  0.22483348, -0.05811155, ..., -0.1859753 ,
         0.31386709,  0.2085023 ]])

In [17]:
cols = pre1.get_feature_names_out()
cols

array(['num__MSSubClass', 'num__LotFrontage', 'num__LotArea',
       'num__OverallQual', 'num__OverallCond', 'num__YearBuilt',
       'num__YearRemodAdd', 'num__MasVnrArea', 'num__BsmtFinSF1',
       'num__BsmtFinSF2', 'num__BsmtUnfSF', 'num__TotalBsmtSF',
       'num__1stFlrSF', 'num__2ndFlrSF', 'num__LowQualFinSF',
       'num__GrLivArea', 'num__BsmtFullBath', 'num__BsmtHalfBath',
       'num__FullBath', 'num__HalfBath', 'num__BedroomAbvGr',
       'num__KitchenAbvGr', 'num__TotRmsAbvGrd', 'num__Fireplaces',
       'num__GarageYrBlt', 'num__GarageCars', 'num__GarageArea',
       'num__WoodDeckSF', 'num__OpenPorchSF', 'num__EnclosedPorch',
       'num__3SsnPorch', 'num__ScreenPorch', 'num__PoolArea',
       'num__MiscVal', 'num__MoSold', 'num__YrSold', 'cat__MSZoning',
       'cat__Street', 'cat__Alley', 'cat__LotShape', 'cat__LandContour',
       'cat__Utilities', 'cat__LotConfig', 'cat__LandSlope',
       'cat__Neighborhood', 'cat__Condition1', 'cat__Condition2',
       'cat__BldgTy

In [18]:
X_pre = pd.DataFrame(X_pre, columns=cols)
X_pre.head()

num__MSSubClass  num__LotFrontage  num__LotArea  num__OverallQual  \
0         0.073375         -0.229372     -0.207142          0.651479   
1        -0.872563          0.451936     -0.091886         -0.071836   
2         0.073375         -0.093110      0.073480          0.651479   
3         0.309859         -0.456474     -0.096897          0.651479   
4         0.073375          0.633618      0.375148          1.374795   

   num__OverallCond  num__YearBuilt  num__YearRemodAdd  num__MasVnrArea  \
0         -0.517200        1.050994           0.878668         0.511418   
1          2.179628        0.156734          -0.429577        -0.574410   
2         -0.517200        0.984752           0.830215         0.323060   
3         -0.517200       -1.863632          -0.720298        -0.574410   
4         -0.517200        0.951632           0.733308         1.364570   

   num__BsmtFinSF1  num__BsmtFinSF2  ...  cat__GarageType  cat__GarageFinish  \
0         0.575425        -0.288653  ...        -0.768736           0.103495   
1         1.171992        -0.288653  ...        -0.768736           0.103495   
2         0.092907        -0.288653  ...        -0.768736           0.103495   
3        -0.499274        -0.288653  ...         1.301075           0.942959   
4         0.463568        -0.288653  ...        -0.768736           0.103495   

   cat__GarageQual  cat__GarageCond  cat__PavedDrive  cat__PoolQC  cat__Fence  \
0         0.315602         0.295349         0.289745     0.063305    0.457447   
1         0.315602         0.295349         0.289745     0.063305    0.457447   
2         0.315602         0.295349         0.289745     0.063305    0.457447   
3         0.315602         0.295349         0.289745     0.063305    0.457447   
4         0.315602         0.295349         0.289745     0.063305    0.457447   

   cat__MiscFeature  cat__SaleType  cat__SaleCondition  
0         -0.185975       0.313867            0.208502  
1         -0.185975       0.313867            0.208502  
2         -0.185975       0.313867            0.208502  
3         -0.185975       0.313867           -3.426284  
4         -0.185975       0.313867            0.208502  

[5 rows x 79 columns]

### Feature Selection

In [19]:
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.linear_model import LinearRegression

In [20]:
lr = LinearRegression()
sel = SequentialFeatureSelector(lr, direction='backward', n_jobs=-1)
sel_features = sel.fit_transform(X_pre, Y)

In [21]:
imp_cols = sel.get_feature_names_out()
imp_cols

array(['num__MSSubClass', 'num__LotArea', 'num__OverallQual',
       'num__OverallCond', 'num__YearBuilt', 'num__MasVnrArea',
       'num__BsmtFinSF1', 'num__BsmtFinSF2', 'num__BsmtUnfSF',
       'num__TotalBsmtSF', 'num__GrLivArea', 'num__BsmtFullBath',
       'num__KitchenAbvGr', 'num__TotRmsAbvGrd', 'num__Fireplaces',
       'num__GarageCars', 'num__WoodDeckSF', 'num__ScreenPorch',
       'num__PoolArea', 'num__YrSold', 'cat__MSZoning', 'cat__Street',
       'cat__Neighborhood', 'cat__BldgType', 'cat__HouseStyle',
       'cat__RoofMatl', 'cat__Exterior1st', 'cat__ExterQual',
       'cat__BsmtQual', 'cat__BsmtCond', 'cat__BsmtExposure',
       'cat__HeatingQC', 'cat__KitchenQual', 'cat__Functional',
       'cat__FireplaceQu', 'cat__GarageFinish', 'cat__GarageCond',
       'cat__Fence', 'cat__MiscFeature', 'cat__SaleCondition'],
      dtype=object)

In [22]:
sel_cols = []
for i in imp_cols:
    s = i.split('__')[1]
    sel_cols.append(s)

In [23]:
sel_cols

['MSSubClass',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 'GrLivArea',
 'BsmtFullBath',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageCars',
 'WoodDeckSF',
 'ScreenPorch',
 'PoolArea',
 'YrSold',
 'MSZoning',
 'Street',
 'Neighborhood',
 'BldgType',
 'HouseStyle',
 'RoofMatl',
 'Exterior1st',
 'ExterQual',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'HeatingQC',
 'KitchenQual',
 'Functional',
 'FireplaceQu',
 'GarageFinish',
 'GarageCond',
 'Fence',
 'MiscFeature',
 'SaleCondition']

In [24]:
X_sel = X[sel_cols]
X_sel.head()

MSSubClass  LotArea  OverallQual  OverallCond  YearBuilt  MasVnrArea  \
0          60     8450            7            5       2003       196.0   
1          20     9600            6            8       1976         0.0   
2          60    11250            7            5       2001       162.0   
3          70     9550            7            5       1915         0.0   
4          60    14260            8            5       2000       350.0   

   BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  ...  BsmtExposure  \
0         706           0        150          856  ...            No   
1         978           0        284         1262  ...            Gd   
2         486           0        434          920  ...            Mn   
3         216           0        540          756  ...            No   
4         655           0        490         1145  ...            Av   

   HeatingQC  KitchenQual  Functional  FireplaceQu  GarageFinish  GarageCond  \
0         Ex           Gd         Typ          NaN           RFn          TA   
1         Ex           TA         Typ           TA           RFn          TA   
2         Ex           Gd         Typ           TA           RFn          TA   
3         Gd           Gd         Typ           Gd           Unf          TA   
4         Ex           Gd         Typ           TA           RFn          TA   

   Fence  MiscFeature  SaleCondition  
0    NaN          NaN         Normal  
1    NaN          NaN         Normal  
2    NaN          NaN         Normal  
3    NaN          NaN        Abnorml  
4    NaN          NaN         Normal  

[5 rows x 40 columns]

### Cat, Con sep

In [25]:
cat_sel, con_sel = catconsep(X_sel)

In [26]:
cat_sel

['MSZoning',
 'Street',
 'Neighborhood',
 'BldgType',
 'HouseStyle',
 'RoofMatl',
 'Exterior1st',
 'ExterQual',
 'BsmtQual',
 'BsmtCond',
 'BsmtExposure',
 'HeatingQC',
 'KitchenQual',
 'Functional',
 'FireplaceQu',
 'GarageFinish',
 'GarageCond',
 'Fence',
 'MiscFeature',
 'SaleCondition']

In [27]:
con_sel

['MSSubClass',
 'LotArea',
 'OverallQual',
 'OverallCond',
 'YearBuilt',
 'MasVnrArea',
 'BsmtFinSF1',
 'BsmtFinSF2',
 'BsmtUnfSF',
 'TotalBsmtSF',
 'GrLivArea',
 'BsmtFullBath',
 'KitchenAbvGr',
 'TotRmsAbvGrd',
 'Fireplaces',
 'GarageCars',
 'WoodDeckSF',
 'ScreenPorch',
 'PoolArea',
 'YrSold']

### Create a final pipeline
Categorical - OneHotEncoded

In [28]:
from sklearn.preprocessing import OneHotEncoder

In [29]:
num_pipe2 = Pipeline(steps=[('imputer',SimpleImputer(strategy='mean')),
                            ('scaler',StandardScaler())])

cat_pipe2 = Pipeline(steps=[('imputer',SimpleImputer(strategy='constant', fill_value='NotAvailable')),
                            ('OHE',OneHotEncoder(handle_unknown='ignore'))])

pre2 = ColumnTransformer([('num',num_pipe2,con_sel),
                          ('cat',cat_pipe2,cat_sel)])

In [30]:
X_sel_pre = pre2.fit_transform(X_sel).toarray()
X_sel_pre

array([[ 0.07337496, -0.20714171,  0.65147924, ...,  0.        ,
         1.        ,  0.        ],
       [-0.87256276, -0.09188637, -0.07183611, ...,  0.        ,
         1.        ,  0.        ],
       [ 0.07337496,  0.07347998,  0.65147924, ...,  0.        ,
         1.        ,  0.        ],
       ...,
       [ 0.30985939, -0.14781027,  0.65147924, ...,  0.        ,
         1.        ,  0.        ],
       [-0.87256276, -0.08016039, -0.79515147, ...,  0.        ,
         1.        ,  0.        ],
       [-0.87256276, -0.05811155, -0.79515147, ...,  0.        ,
         1.        ,  0.        ]])

In [31]:
final_cols = pre2.get_feature_names_out()
final_cols

array(['num__MSSubClass', 'num__LotArea', 'num__OverallQual',
       'num__OverallCond', 'num__YearBuilt', 'num__MasVnrArea',
       'num__BsmtFinSF1', 'num__BsmtFinSF2', 'num__BsmtUnfSF',
       'num__TotalBsmtSF', 'num__GrLivArea', 'num__BsmtFullBath',
       'num__KitchenAbvGr', 'num__TotRmsAbvGrd', 'num__Fireplaces',
       'num__GarageCars', 'num__WoodDeckSF', 'num__ScreenPorch',
       'num__PoolArea', 'num__YrSold', 'cat__MSZoning_C (all)',
       'cat__MSZoning_FV', 'cat__MSZoning_RH', 'cat__MSZoning_RL',
       'cat__MSZoning_RM', 'cat__Street_Grvl', 'cat__Street_Pave',
       'cat__Neighborhood_Blmngtn', 'cat__Neighborhood_Blueste',
       'cat__Neighborhood_BrDale', 'cat__Neighborhood_BrkSide',
       'cat__Neighborhood_ClearCr', 'cat__Neighborhood_CollgCr',
       'cat__Neighborhood_Crawfor', 'cat__Neighborhood_Edwards',
       'cat__Neighborhood_Gilbert', 'cat__Neighborhood_IDOTRR',
       'cat__Neighborhood_MeadowV', 'cat__Neighborhood_Mitchel',
       'cat__Neighborhood_

In [32]:
X_sel_pre = pd.DataFrame(X_sel_pre,columns=final_cols)
X_sel_pre.head()

num__MSSubClass  num__LotArea  num__OverallQual  num__OverallCond  \
0         0.073375     -0.207142          0.651479         -0.517200   
1        -0.872563     -0.091886         -0.071836          2.179628   
2         0.073375      0.073480          0.651479         -0.517200   
3         0.309859     -0.096897          0.651479         -0.517200   
4         0.073375      0.375148          1.374795         -0.517200   

   num__YearBuilt  num__MasVnrArea  num__BsmtFinSF1  num__BsmtFinSF2  \
0        1.050994         0.511418         0.575425        -0.288653   
1        0.156734        -0.574410         1.171992        -0.288653   
2        0.984752         0.323060         0.092907        -0.288653   
3       -1.863632        -0.574410        -0.499274        -0.288653   
4        0.951632         1.364570         0.463568        -0.288653   

   num__BsmtUnfSF  num__TotalBsmtSF  ...  cat__MiscFeature_NotAvailable  \
0       -0.944591         -0.459303  ...                            1.0   
1       -0.641228          0.466465  ...                            1.0   
2       -0.301643         -0.313369  ...                            1.0   
3       -0.061670         -0.687324  ...                            1.0   
4       -0.174865          0.199680  ...                            1.0   

   cat__MiscFeature_Othr  cat__MiscFeature_Shed  cat__MiscFeature_TenC  \
0                    0.0                    0.0                    0.0   
1                    0.0                    0.0                    0.0   
2                    0.0                    0.0                    0.0   
3                    0.0                    0.0                    0.0   
4                    0.0                    0.0                    0.0   

   cat__SaleCondition_Abnorml  cat__SaleCondition_AdjLand  \
0                         0.0                         0.0   
1                         0.0                         0.0   
2                         0.0                         0.0   
3                         1.0                         0.0   
4                         0.0                         0.0   

   cat__SaleCondition_Alloca  cat__SaleCondition_Family  \
0                        0.0                        0.0   
1                        0.0                        0.0   
2                        0.0                        0.0   
3                        0.0                        0.0   
4                        0.0                        0.0   

   cat__SaleCondition_Normal  cat__SaleCondition_Partial  
0                        1.0                         0.0  
1                        1.0                         0.0  
2                        1.0                         0.0  
3                        0.0                         0.0  
4                        1.0                         0.0  

[5 rows x 155 columns]

### Train Test split
Train : 80% Test : 20%

In [33]:
from sklearn.model_selection import train_test_split
xtrain, xtest, ytrain, ytest = train_test_split(X_sel_pre, Y, test_size=0.2, random_state=21)

### Create a base model

In [34]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(xtrain, ytrain)

LinearRegression()

In [35]:
lr.score(xtrain, ytrain)

0.9273466880156978

In [36]:
lr.score(xtest, ytest)

0.8203751914348976

In [37]:
from PM6 import evaluate_model
evaluate_model(xtrain, ytrain, xtest, ytest, lr)

Training Results :

MSE : 449645936.75
RMSE: 21204.86
MAE : 14412.24
R2  : 0.9273


Testing Results :

MSE : 1217373367.55
RMSE: 34890.88
MAE : 16729.86
R2  : 0.8204


### Cross validated results

In [38]:
from sklearn.model_selection import cross_val_score

In [39]:
r2_scores = cross_val_score(lr, xtrain, ytrain, cv=5, scoring='r2')
r2_scores

array([-5.80563813e+21, -2.63981335e+19,  9.09944368e-01, -7.48054154e+17,
       -1.52974861e+17])

In [40]:
r2_cv = r2_scores.mean()
r2_cv

-1.166587457845499e+21

In [41]:
mse_scores = cross_val_score(lr, xtrain, ytrain, cv=5, scoring='neg_mean_squared_error')
mse_scores

array([-3.75094788e+31, -1.56734495e+29, -4.69922277e+08, -5.38434755e+27,
       -9.23544456e+26])

In [42]:
mse_cv = -mse_scores.mean()
mse_cv

7.534504241303522e+30

In [43]:
rsme_scores = cross_val_score(lr, xtrain, ytrain, cv=5, scoring='neg_root_mean_squared_error')
rsme_scores

array([-6.12449825e+15, -3.95897076e+14, -2.16776908e+04, -7.33781135e+13,
       -3.03898742e+13])

In [44]:
rsme_cv = -rsme_scores.mean()
rsme_cv

1324832662815581.2

In [45]:
mae_scores = cross_val_score(lr, xtrain, ytrain, cv=5, scoring='neg_mean_absolute_error')
mae_scores

array([-9.13219632e+14, -4.11063526e+13, -1.62607153e+04, -4.80716007e+12,
       -3.44802722e+12])

In [46]:
mae_cv = -mae_scores.mean()
mae_cv

192516234391500.72

### Final model building

### Ridge model

In [47]:
import numpy as np
alphas = np.arange(0.1, 100, 0.1)
alphas

array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
        7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
        8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
       10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. ,
       11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1,
       12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2,
       13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3,
       14.4, 14.5, 14.6, 14.7, 14.8, 14.9, 15. , 15

In [48]:
params = {'alpha':alphas}
params

{'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
         1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
         2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
         3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
         4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
         5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
         6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7.4,  7.5,  7.6,  7.7,
         7.8,  7.9,  8. ,  8.1,  8.2,  8.3,  8.4,  8.5,  8.6,  8.7,  8.8,
         8.9,  9. ,  9.1,  9.2,  9.3,  9.4,  9.5,  9.6,  9.7,  9.8,  9.9,
        10. , 10.1, 10.2, 10.3, 10.4, 10.5, 10.6, 10.7, 10.8, 10.9, 11. ,
        11.1, 11.2, 11.3, 11.4, 11.5, 11.6, 11.7, 11.8, 11.9, 12. , 12.1,
        12.2, 12.3, 12.4, 12.5, 12.6, 12.7, 12.8, 12.9, 13. , 13.1, 13.2,
        13.3, 13.4, 13.5, 13.6, 13.7, 13.8, 13.9, 14. , 14.1, 14.2, 14.3,
        14.4, 14.5, 14.6, 14.

In [49]:
from sklearn.linear_model import Ridge
from sklearn.model_selection import GridSearchCV

In [50]:
ridge = Ridge()
gscv = GridSearchCV(ridge, param_grid=params, cv=5, scoring='neg_mean_squared_error')
gscv.fit(xtrain, ytrain)

GridSearchCV(cv=5, estimator=Ridge(),
             param_grid={'alpha': array([ 0.1,  0.2,  0.3,  0.4,  0.5,  0.6,  0.7,  0.8,  0.9,  1. ,  1.1,
        1.2,  1.3,  1.4,  1.5,  1.6,  1.7,  1.8,  1.9,  2. ,  2.1,  2.2,
        2.3,  2.4,  2.5,  2.6,  2.7,  2.8,  2.9,  3. ,  3.1,  3.2,  3.3,
        3.4,  3.5,  3.6,  3.7,  3.8,  3.9,  4. ,  4.1,  4.2,  4.3,  4.4,
        4.5,  4.6,  4.7,  4.8,  4.9,  5. ,  5.1,  5.2,  5.3,  5.4,  5.5,
        5.6,  5.7,  5.8,  5.9,  6. ,  6.1,  6.2,  6.3,  6.4,  6.5,  6.6,
        6.7,  6.8,  6.9,  7. ,  7.1,  7.2,  7.3,  7....
       93.6, 93.7, 93.8, 93.9, 94. , 94.1, 94.2, 94.3, 94.4, 94.5, 94.6,
       94.7, 94.8, 94.9, 95. , 95.1, 95.2, 95.3, 95.4, 95.5, 95.6, 95.7,
       95.8, 95.9, 96. , 96.1, 96.2, 96.3, 96.4, 96.5, 96.6, 96.7, 96.8,
       96.9, 97. , 97.1, 97.2, 97.3, 97.4, 97.5, 97.6, 97.7, 97.8, 97.9,
       98. , 98.1, 98.2, 98.3, 98.4, 98.5, 98.6, 98.7, 98.8, 98.9, 99. ,
       99.1, 99.2, 99.3, 99.4, 99.5, 99.6, 99.7, 99.8, 99.9])},
             scoring='neg_mean_squared_error')

In [51]:
gscv.best_params_

{'alpha': 19.700000000000003}

In [52]:
gscv.best_score_

-1045136353.6879857

In [53]:
best_ridge = gscv.best_estimator_
best_ridge

Ridge(alpha=19.700000000000003)

In [54]:
best_ridge.score(xtrain, ytrain)

0.8845114791172876

In [55]:
best_ridge.score(xtest, ytest)

0.8356167597249187

In [56]:
r2_scores_r = cross_val_score(best_ridge, xtrain, ytrain, cv=5, scoring='r2')
r2_scores_r

array([0.60946801, 0.85025648, 0.90214539, 0.8785968 , 0.92894538])

In [57]:
r2_scores_r.mean()

0.833882413976686

In [58]:
evaluate_model(xtrain, ytrain, xtest, ytest, best_ridge)

Training Results :

MSE : 714749854.32
RMSE: 26734.81
MAE : 15857.47
R2  : 0.8845


Testing Results :

MSE : 1114076504.13
RMSE: 33377.78
MAE : 16944.56
R2  : 0.8356


### Lasso model

In [59]:
import numpy as np
alphas2 = np.arange(100, 200, 0.1)
alphas2

array([100. , 100.1, 100.2, 100.3, 100.4, 100.5, 100.6, 100.7, 100.8,
       100.9, 101. , 101.1, 101.2, 101.3, 101.4, 101.5, 101.6, 101.7,
       101.8, 101.9, 102. , 102.1, 102.2, 102.3, 102.4, 102.5, 102.6,
       102.7, 102.8, 102.9, 103. , 103.1, 103.2, 103.3, 103.4, 103.5,
       103.6, 103.7, 103.8, 103.9, 104. , 104.1, 104.2, 104.3, 104.4,
       104.5, 104.6, 104.7, 104.8, 104.9, 105. , 105.1, 105.2, 105.3,
       105.4, 105.5, 105.6, 105.7, 105.8, 105.9, 106. , 106.1, 106.2,
       106.3, 106.4, 106.5, 106.6, 106.7, 106.8, 106.9, 107. , 107.1,
       107.2, 107.3, 107.4, 107.5, 107.6, 107.7, 107.8, 107.9, 108. ,
       108.1, 108.2, 108.3, 108.4, 108.5, 108.6, 108.7, 108.8, 108.9,
       109. , 109.1, 109.2, 109.3, 109.4, 109.5, 109.6, 109.7, 109.8,
       109.9, 110. , 110.1, 110.2, 110.3, 110.4, 110.5, 110.6, 110.7,
       110.8, 110.9, 111. , 111.1, 111.2, 111.3, 111.4, 111.5, 111.6,
       111.7, 111.8, 111.9, 112. , 112.1, 112.2, 112.3, 112.4, 112.5,
       112.6, 112.7,

In [60]:
params2 = {'alpha':alphas2}
params2

{'alpha': array([100. , 100.1, 100.2, 100.3, 100.4, 100.5, 100.6, 100.7, 100.8,
        100.9, 101. , 101.1, 101.2, 101.3, 101.4, 101.5, 101.6, 101.7,
        101.8, 101.9, 102. , 102.1, 102.2, 102.3, 102.4, 102.5, 102.6,
        102.7, 102.8, 102.9, 103. , 103.1, 103.2, 103.3, 103.4, 103.5,
        103.6, 103.7, 103.8, 103.9, 104. , 104.1, 104.2, 104.3, 104.4,
        104.5, 104.6, 104.7, 104.8, 104.9, 105. , 105.1, 105.2, 105.3,
        105.4, 105.5, 105.6, 105.7, 105.8, 105.9, 106. , 106.1, 106.2,
        106.3, 106.4, 106.5, 106.6, 106.7, 106.8, 106.9, 107. , 107.1,
        107.2, 107.3, 107.4, 107.5, 107.6, 107.7, 107.8, 107.9, 108. ,
        108.1, 108.2, 108.3, 108.4, 108.5, 108.6, 108.7, 108.8, 108.9,
        109. , 109.1, 109.2, 109.3, 109.4, 109.5, 109.6, 109.7, 109.8,
        109.9, 110. , 110.1, 110.2, 110.3, 110.4, 110.5, 110.6, 110.7,
        110.8, 110.9, 111. , 111.1, 111.2, 111.3, 111.4, 111.5, 111.6,
        111.7, 111.8, 111.9, 112. , 112.1, 112.2, 112.3, 112.4, 112.

In [61]:
from sklearn.linear_model import Lasso
lasso = Lasso()
gscv2 = GridSearchCV(lasso, param_grid=params2, cv=5, scoring='neg_mean_squared_error')
gscv2.fit(xtrain, ytrain)

GridSearchCV(cv=5, estimator=Lasso(),
             param_grid={'alpha': array([100. , 100.1, 100.2, 100.3, 100.4, 100.5, 100.6, 100.7, 100.8,
       100.9, 101. , 101.1, 101.2, 101.3, 101.4, 101.5, 101.6, 101.7,
       101.8, 101.9, 102. , 102.1, 102.2, 102.3, 102.4, 102.5, 102.6,
       102.7, 102.8, 102.9, 103. , 103.1, 103.2, 103.3, 103.4, 103.5,
       103.6, 103.7, 103.8, 103.9, 104. , 104.1, 104.2, 104.3, 104.4,
       104.5, 104.6, 104.7, 104....
       194.5, 194.6, 194.7, 194.8, 194.9, 195. , 195.1, 195.2, 195.3,
       195.4, 195.5, 195.6, 195.7, 195.8, 195.9, 196. , 196.1, 196.2,
       196.3, 196.4, 196.5, 196.6, 196.7, 196.8, 196.9, 197. , 197.1,
       197.2, 197.3, 197.4, 197.5, 197.6, 197.7, 197.8, 197.9, 198. ,
       198.1, 198.2, 198.3, 198.4, 198.5, 198.6, 198.7, 198.8, 198.9,
       199. , 199.1, 199.2, 199.3, 199.4, 199.5, 199.6, 199.7, 199.8,
       199.9])},
             scoring='neg_mean_squared_error')

In [62]:
gscv2.best_params_

{'alpha': 109.89999999999944}

In [63]:
gscv2.best_score_

-961670609.3397844

In [64]:
best_lasso = gscv2.best_estimator_
best_lasso

Lasso(alpha=109.89999999999944)

In [65]:
best_lasso.score(xtrain, ytrain)

0.9166186729474692

In [66]:
best_lasso.score(xtest, ytest)

0.8287661604119128

In [67]:
r2_scores_l = cross_val_score(best_lasso, xtrain, ytrain, cv=5, scoring='r2')
r2_scores_l

array([0.62532587, 0.86937292, 0.91391516, 0.89181777, 0.93636506])

In [68]:
r2_scores_l.mean()

0.8473593573306395

In [69]:
evaluate_model(xtrain, ytrain, xtest, ytest, best_lasso)

Training Results :

MSE : 516040823.00
RMSE: 22716.53
MAE : 15015.26
R2  : 0.9166


Testing Results :

MSE : 1160505152.95
RMSE: 34066.19
MAE : 16414.51
R2  : 0.8288


### On basis of testing data
1. Ridge model R2 Test : 0.8356
2. Lasso model R2 Test : 0.8288
    
Select Ridge model as final model

In [70]:
ypred_tr = best_ridge.predict(xtrain)
ypred_ts = best_ridge.predict(xtest)

In [71]:
ypred_tr[0:5]

array([[ 72268.50219659],
       [105975.27784309],
       [161786.0741962 ],
       [218569.09538921],
       [216338.9134092 ]])

In [72]:
ytrain.head()

SalePrice
710       52000
1098     128000
1286     143000
992      187000
631      209500

In [73]:
ypred_ts[0:5]

array([[168547.13540343],
       [202051.8076491 ],
       [233959.77540144],
       [218227.99380292],
       [ 79080.15896399]])

In [74]:
ytest.head()

SalePrice
880      157000
605      205000
1166     245350
216      210000
970      135000

### Predicting the sales price

In [75]:
xnew = pd.read_csv('sample_set.csv')
xnew

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave   NaN      Reg   
1     1462          20       RL         81.0    14267   Pave   NaN      IR1   
2     1463          60       RL         74.0    13830   Pave   NaN      IR1   
3     1464          60       RL         78.0     9978   Pave   NaN      IR1   
4     1465         120       RL         43.0     5005   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1454  2915         160       RM         21.0     1936   Pave   NaN      Reg   
1455  2916         160       RM         21.0     1894   Pave   NaN      Reg   
1456  2917          20       RL        160.0    20000   Pave   NaN      Reg   
1457  2918          85       RL         62.0    10441   Pave   NaN      Reg   
1458  2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence  \
0            Lvl    AllPub  ...         120        0    NaN  MnPrv   
1            Lvl    AllPub  ...           0        0    NaN    NaN   
2            Lvl    AllPub  ...           0        0    NaN  MnPrv   
3            Lvl    AllPub  ...           0        0    NaN    NaN   
4            HLS    AllPub  ...         144        0    NaN    NaN   
...          ...       ...  ...         ...      ...    ...    ...   
1454         Lvl    AllPub  ...           0        0    NaN    NaN   
1455         Lvl    AllPub  ...           0        0    NaN    NaN   
1456         Lvl    AllPub  ...           0        0    NaN    NaN   
1457         Lvl    AllPub  ...           0        0    NaN  MnPrv   
1458         Lvl    AllPub  ...           0        0    NaN    NaN   

     MiscFeature MiscVal MoSold  YrSold  SaleType  SaleCondition  
0            NaN       0      6    2010        WD         Normal  
1           Gar2   12500      6    2010        WD         Normal  
2            NaN       0      3    2010        WD         Normal  
3            NaN       0      6    2010        WD         Normal  
4            NaN       0      1    2010        WD         Normal  
...          ...     ...    ...     ...       ...            ...  
1454         NaN       0      6    2006        WD         Normal  
1455         NaN       0      4    2006        WD        Abnorml  
1456         NaN       0      9    2006        WD        Abnorml  
1457        Shed     700      7    2006        WD         Normal  
1458         NaN       0     11    2006        WD         Normal  

[1459 rows x 80 columns]

In [76]:
xnew_pre = pre2.transform(xnew).toarray()
xnew_pre

array([[-0.87256276,  0.11076257, -0.79515147, ...,  0.        ,
         1.        ,  0.        ],
       [-0.87256276,  0.37584985, -0.07183611, ...,  0.        ,
         1.        ,  0.        ],
       [ 0.07337496,  0.33205282, -0.79515147, ...,  0.        ,
         1.        ,  0.        ],
       ...,
       [-0.87256276,  0.95042275, -0.79515147, ...,  0.        ,
         0.        ,  0.        ],
       [ 0.66458604, -0.00759964, -0.79515147, ...,  0.        ,
         1.        ,  0.        ],
       [ 0.07337496, -0.08918038,  0.65147924, ...,  0.        ,
         1.        ,  0.        ]])

In [77]:
xnew_pre = pd.DataFrame(xnew_pre, columns=final_cols)
xnew_pre

num__MSSubClass  num__LotArea  num__OverallQual  num__OverallCond  \
0           -0.872563      0.110763         -0.795151          0.381743   
1           -0.872563      0.375850         -0.071836          0.381743   
2            0.073375      0.332053         -0.795151         -0.517200   
3            0.073375     -0.054002         -0.071836          0.381743   
4            1.492282     -0.552407          1.374795         -0.517200   
...               ...           ...               ...               ...   
1454         2.438219     -0.859988         -1.518467          1.280685   
1455         2.438219     -0.864197         -1.518467         -0.517200   
1456        -0.872563      0.950423         -0.795151          1.280685   
1457         0.664586     -0.007600         -0.795151         -0.517200   
1458         0.073375     -0.089180          0.651479         -0.517200   

      num__YearBuilt  num__MasVnrArea  num__BsmtFinSF1  num__BsmtFinSF2  \
0          -0.340077        -0.574410         0.053428         0.604293   
1          -0.439440         0.023903         1.051363        -0.288653   
2           0.852269        -0.574410         0.761852        -0.288653   
3           0.885390        -0.463612         0.347326        -0.288653   
4           0.686666        -0.574410        -0.396190        -0.288653   
...              ...              ...              ...              ...   
1454       -0.041991        -0.574410        -0.973018        -0.288653   
1455       -0.041991        -0.574410        -0.420316        -0.288653   
1456       -0.373198        -0.574410         1.711535        -0.288653   
1457        0.686666        -0.574410        -0.233889        -0.288653   
1458        0.719786        -0.053656         0.689474        -0.288653   

      num__BsmtUnfSF  num__TotalBsmtSF  ...  cat__MiscFeature_NotAvailable  \
0          -0.672923         -0.400017  ...                            1.0   
1          -0.365032          0.619239  ...                            0.0   
2          -0.974021         -0.295127  ...                            1.0   
3          -0.550672         -0.299687  ...                            1.0   
4           1.018211          0.507509  ...                            1.0   
...              ...               ...  ...                            ...   
1454       -0.048086         -1.166169  ...                            1.0   
1455       -0.618589         -1.166169  ...                            1.0   
1456       -1.284176          0.379817  ...                            1.0   
1457        0.017567         -0.331610  ...                            0.0   
1458       -0.745367         -0.140072  ...                            1.0   

      cat__MiscFeature_Othr  cat__MiscFeature_Shed  cat__MiscFeature_TenC  \
0                       0.0                    0.0                    0.0   
1                       0.0                    0.0                    0.0   
2                       0.0                    0.0                    0.0   
3                       0.0                    0.0                    0.0   
4                       0.0                    0.0                    0.0   
...                     ...                    ...                    ...   
1454                    0.0                    0.0                    0.0   
1455                    0.0                    0.0                    0.0   
1456                    0.0                    0.0                    0.0   
1457                    0.0                    1.0                    0.0   
1458                    0.0                    0.0                    0.0   

      cat__SaleCondition_Abnorml  cat__SaleCondition_AdjLand  \
0                            0.0                         0.0   
1                            0.0                         0.0   
2                            0.0                         0.0   
3                            0.0                         0.0   
4                     

In [78]:
preds = best_ridge.predict(xnew_pre)
preds

array([[118007.04578693],
       [151190.41082164],
       [170080.67332129],
       ...,
       [175480.0219072 ],
       [105808.59563773],
       [229901.86013185]])

In [79]:
xnew['SalePrice_pred']=preds
xnew

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave   NaN      Reg   
1     1462          20       RL         81.0    14267   Pave   NaN      IR1   
2     1463          60       RL         74.0    13830   Pave   NaN      IR1   
3     1464          60       RL         78.0     9978   Pave   NaN      IR1   
4     1465         120       RL         43.0     5005   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1454  2915         160       RM         21.0     1936   Pave   NaN      Reg   
1455  2916         160       RM         21.0     1894   Pave   NaN      Reg   
1456  2917          20       RL        160.0    20000   Pave   NaN      Reg   
1457  2918          85       RL         62.0    10441   Pave   NaN      Reg   
1458  2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities  ... PoolArea PoolQC  Fence MiscFeature MiscVal  \
0            Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
1            Lvl    AllPub  ...        0    NaN    NaN        Gar2   12500   
2            Lvl    AllPub  ...        0    NaN  MnPrv         NaN       0   
3            Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
4            HLS    AllPub  ...        0    NaN    NaN         NaN       0   
...          ...       ...  ...      ...    ...    ...         ...     ...   
1454         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1455         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1456         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   
1457         Lvl    AllPub  ...        0    NaN  MnPrv        Shed     700   
1458         Lvl    AllPub  ...        0    NaN    NaN         NaN       0   

     MoSold YrSold  SaleType  SaleCondition  SalePrice_pred  
0         6   2010        WD         Normal   118007.045787  
1         6   2010        WD         Normal   151190.410822  
2         3   2010        WD         Normal   170080.673321  
3         6   2010        WD         Normal   188329.697767  
4         1   2010        WD         Normal   190394.155482  
...     ...    ...       ...            ...             ...  
1454      6   2006        WD         Normal    79079.143720  
1455      4   2006        WD        Abnorml    72575.096577  
1456      9   2006        WD        Abnorml   175480.021907  
1457      7   2006        WD         Normal   105808.595638  
1458     11   2006        WD         Normal   229901.860132  

[1459 rows x 81 columns]

In [80]:
A = xnew[['Id','SalePrice_pred']]
A

Id  SalePrice_pred
0     1461   118007.045787
1     1462   151190.410822
2     1463   170080.673321
3     1464   188329.697767
4     1465   190394.155482
...    ...             ...
1454  2915    79079.143720
1455  2916    72575.096577
1456  2917   175480.021907
1457  2918   105808.595638
1458  2919   229901.860132

[1459 rows x 2 columns]

In [81]:
A.to_csv('SalesPrice.csv',index=False)